lib/matplotlib/legend.py

```diff
@@ -763,28 +764,35 @@ def _init_legend_box(self, handles, labels, markerfirst=True):
         for orig_handle, lab in zip(handles, labels):
             handler = self.get_legend_handler(legend_handler_map, orig_handle)
             if handler is None:
-                _api.warn_external(
-                    "Legend does not support {!r} instances.\nA proxy artist "
-                    "may be used instead.\nSee: "
-                    "https://matplotlib.org/users/legend_guide.html"
-                    "#creating-artists-specifically-for-adding-to-the-legend-"
-                    "aka-proxy-artists".format(orig_handle))
-                # We don't have a handle for this artist, so we just defer
-                # to None.
-                handle_list.append(None)
-            else:
-                textbox = TextArea(lab, textprops=label_prop,
-                                   multilinebaseline=True)
-                handlebox = DrawingArea(width=self.handlelength * fontsize,
-                                        height=height,
-                                        xdescent=0., ydescent=descent)
-
-                text_list.append(textbox._text)
-                # Create the artist for the legend which represents the
-                # original artist/handle.
-                handle_list.append(handler.legend_artist(self, orig_handle,
-                                                         fontsize, handlebox))
-                handles_and_labels.append((handlebox, textbox))
+                if isinstance(orig_handle, Text):
+                    _api.warn_external(
+                            "Legend does not support handles for Text "
+                            "instances.\nSee: https://matplotlib.org/stable/"
+                            "tutorials/intermediate/legend_guide.html"
+                            "#implementing-a-custom-legend-handler")
+                    continue
+                else:
```

Comment:
This is going to provide switch _every_ Artists that is not `Text` to fallback to `Rectangle`!  That seems a bit over-broad and I am not sure the `Rectangle` handler will be able to extract the information it needs from an arbitrary `Artist`.

Reply:
The only `Artists` that are iterated over by default are: `Line2D`, `Patch`, `Collection`, and `Text`.

`Line2D` and (some?) `Collection` are in the `_default_handler_map`.

`Text` cannot be put in `_default_handler_map` unless a handler is written for it, and the solution for the linked issue was scoped to issuing a warning.

Before this PR, any `Patch` mapped to a `HandlerPatch`, which was a `Rectangle`. I've made that behavior explicit.

Reply:
This is called from https://github.com/matplotlib/matplotlib/blob/3e32cc7e88d6d583942335dff4c83b4b7dd13ae9/lib/matplotlib/legend.py#L493-L494 which are lightly processed at https://github.com/matplotlib/matplotlib/blob/3e32cc7e88d6d583942335dff4c83b4b7dd13ae9/lib/matplotlib/legend.py#L401-L411 but otherwise come in from https://github.com/matplotlib/matplotlib/blob/3e32cc7e88d6d583942335dff4c83b4b7dd13ae9/lib/matplotlib/legend.py#L294-L295 which is in turn called from https://github.com/matplotlib/matplotlib/blob/3e32cc7e88d6d583942335dff4c83b4b7dd13ae9/lib/matplotlib/axes/_axes.py#L299-L305 which has the path https://github.com/matplotlib/matplotlib/blob/3e32cc7e88d6d583942335dff4c83b4b7dd13ae9/lib/matplotlib/legend.py#L1210-L1211 and https://github.com/matplotlib/matplotlib/blob/3e32cc7e88d6d583942335dff4c83b4b7dd13ae9/lib/matplotlib/legend.py#L1240-L1242 that means the user can (eventually) get arbitrary artists into `handles` here so we can not assume that the input types will be limited to what the auto-discovery code picks up.

Reply:
But the arbitrary artist needs a handler for it to be rendered in the legend; can a user pass an arbitrary artist in this way without one? So `handler` would not be `None` here:

https://github.com/matplotlib/matplotlib/blob/7d43704830e92059e40dcdf09561ec0f50bac37a/lib/matplotlib/legend.py#L765

and it wouldn't need to fall back to a `Rectangle`.

Reply:
However if the user _does_make a mistake and pass in an artist that does not have a handler entry we should not then fallback to `Rectangle`, we should maintain the current behavior of warning and passing on the handle.